File: backend_python/app/models/finding.py
from datetime import datetime
from typing import Optional, Literal
from pydantic import BaseModel, Field

class BoundingBox(BaseModel):
    left: str
    top: str
    width: str
    height: str

class FindingModel(BaseModel):
    id: str = Field(alias="_id")  # Custom String primary key (e.g. "IR-4821")
    batchId: str  # Ref to InspectionBatch._id
    productId: str  # Ref to Product._id
    modelRunId: Optional[str] = None  # Ref to ModelRun._id
    findingCode: str
    defectType: str
    severity: Literal["Low", "Medium", "High", "Critical"] = "Low"
    severityScore: float = 0.0
    confidence: float = 0.0
    defectArea: str = "0.0%"
    decision: str = "Review queued"
    boundingBox: Optional[BoundingBox] = None
    segmentationMaskImageId: Optional[str] = None
    gradcamImageId: Optional[str] = None
    boundingBoxImageId: Optional[str] = None
    isFlagged: bool = False
    createdAt: datetime = Field(default_factory=datetime.utcnow)
    updatedAt: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        populate_by_name = True

File: backend/server/models/Finding.ts
import mongoose, { Schema, Document } from "mongoose";

export interface IBoundingBox {
  left: string;
  top: string;
  width: string;
  height: string;
}

export interface IFinding extends Document {
  _id: string; // Primary key (e.g. "IR-4821")
  batchId: string;
  productId: string;
  modelRunId?: string | null;
  findingCode: string;
  defectType: string;
  severity: "Low" | "Medium" | "High" | "Critical";
  severityScore: number;
  confidence: number;
  defectArea?: string;
  decision?: string;
  boundingBox?: IBoundingBox | null;
  segmentationMaskImageId?: string | null;
  gradcamImageId?: string | null;
  boundingBoxImageId?: string | null;
  isFlagged: boolean;
  createdAt: Date;
  updatedAt: Date;
}

const BoundingBoxSchema = new Schema(
  {
    left: { type: String, required: true },
    top: { type: String, required: true },
    width: { type: String, required: true },
    height: { type: String, required: true }
  },
  { _id: false }
);

const FindingSchema = new Schema<IFinding>(
  {
    _id: { type: String, required: true },
    batchId: { type: String, ref: "InspectionBatch", required: true },
    productId: { type: String, ref: "Product", required: true },
    modelRunId: { type: String, ref: "ModelRun", default: null },
    findingCode: { type: String, required: true },
    defectType: { type: String, required: true },
    severity: {
      type: String,
      enum: ["Low", "Medium", "High", "Critical"],
      default: "Low"
    },
    severityScore: { type: Number, default: 0 },
    confidence: { type: Number, default: 0 },
    defectArea: { type: String, default: "0.0%" },
    decision: { type: String, default: "Review queued" },
    boundingBox: { type: BoundingBoxSchema, default: null },
    segmentationMaskImageId: { type: String, ref: "InspectionImage", default: null },
    gradcamImageId: { type: String, ref: "InspectionImage", default: null },
    boundingBoxImageId: { type: String, ref: "InspectionImage", default: null },
    isFlagged: { type: Boolean, default: false }
  },
  {
    timestamps: true,
    _id: false
  }
);

FindingSchema.index({ productId: 1, isFlagged: 1, severity: 1 });
FindingSchema.index({ batchId: 1, isFlagged: 1 });
FindingSchema.index({ findingCode: 1 });

export const Finding =
  mongoose.models.Finding ||
  mongoose.model<IFinding>("Finding", FindingSchema);
